### PR TITLE
fix(TwabRewards): store prize pool ticket address

### PR DESCRIPTION
**Pull request by PierrickGT**
Issue: https://github.com/code-423n4/2021-12-pooltogether-findings/issues/1

### CODE REVIEW

File: contracts/TwabRewards.sol [hunk 1]
```diff
@@ -16,13 +16,17 @@ import "./interfaces/ITwabRewards.sol";
  * In order to calculate user rewards, we use the TWAB (Time-Weighted Average Balance) from the Ticket contract.
  * This way, users simply need to hold their tickets to be eligible to claim rewards.
  * Rewards are calculated based on the average amount of tickets they hold during the epoch duration.
+ * @dev This contract support only one prize pool ticket.
```
Comment (aodhgan) [inline contracts/TwabRewards.sol:19]: nit: supports

File: contracts/interfaces/ITwabRewards.sol [hunk 2]
```diff
@@ -22,7 +22,6 @@ interface ITwabRewards {
      */
     struct Promotion {
```
Comment (asselstine) [inline contracts/interfaces/ITwabRewards.sol:23]: The promotion can be packed more tightly, which will allow larger uints.

Keep in mind `address` is 20 bytes, or 160 bits.

32 bit timestamp overflow in 100 years.  Lots of time yes, but we can remove all worries by making timestamps as large as possible.

Like so:

```
struct Promotion {
  address creator // 160 bits
  IERC20 token // 320 bits
  uint128 startTimestamp // 448 bits
  uint56 epochDuration // 504 bits
  uint8 numberOfEpochs // 512 bits
  uint256 tokensPerEpoch // 768 bits
}
```

total words: 3 (same as before)

**Review by asselstine — CHANGES_REQUESTED** [review]

File: contracts/TwabRewards.sol [hunk 3]
```diff
@@ -16,13 +16,17 @@ import "./interfaces/ITwabRewards.sol";
...
  * @dev This contract does not support the use of fee on transfer tokens.
  */
 contract TwabRewards is ITwabRewards {
     using SafeERC20 for IERC20;
 
     /* ============ Global Variables ============ */
 
+    /// @notice Prize pool ticket for which the promotions are created.
+    ITicket public ticket;
```
[the remark this answers is not in the document]
Comment (PierrickGT) [reply contracts/TwabRewards.sol]: Done: https://github.com/pooltogether/v4-periphery/pull/47/commits/c480f0e3fab1015dcfb267d88b7e1b071d541e87

File: contracts/interfaces/ITwabRewards.sol (hunk 2 above)
Comment (PierrickGT) [reply contracts/interfaces/ITwabRewards.sol:22]: The order of variables matter for packing to work. As stated in this article: https://www.maxgrok.com/posts/one-hundred-and-nineth-post
Keeping this in mind, I propose this packing:
```
    struct Promotion {
        address creator; // 160 bits
        uint64 startTimestamp; // 224 bits
        uint8 numberOfEpochs; // 232 bits -> one word
        uint64 epochDuration; // 64 bits
        IERC20 token; // 224 bits -> one word
        uint256 tokensPerEpoch; // 256 bits -> one word
    }
```

This way, we still have 3 words and we don't have to cast timestamps to uint64 to interact with the `getAverageBalanceBetween` function: https://github.com/pooltogether/v4-periphery/blob/0e2de252f61725f44999016cb769fdc99539b641/contracts/TwabRewards.sol#L361-L372

**Pull request merged**